astropy/units/format/generic.py

```diff
@@ -604,18 +604,55 @@ def _validate_unit(cls, unit: str, detailed_exception: bool = True) -> None:
             if detailed_exception:
                 raise ValueError(
                     f"Unit '{unit}' not supported by the {cls.__name__} standard. "
-                    + did_you_mean_units(
-                        unit,
-                        cls._units,
-                        cls._deprecated_units,
-                        cls._to_decomposed_alternative,
-                    ),
+                    + cls._did_you_mean_units(unit)
                 )
             raise ValueError()
         if unit in cls._deprecated_units:
-            unit_deprecation_warning(
-                unit, cls._units[unit], cls.__name__, cls._to_decomposed_alternative
+            from astropy.units.core import UnitsWarning
```

Comment:
I assume there is a circular import error if put on top?

Reply:
Yes, that would cause circular imports.

I don't see any good reasons why the error and warning definitions have to be in `astropy.units.core`. They could be in a separate file, which would not have to import anything else from `astropy` so it could not cause any import loops. But creating a new file for unit errors and warnings requires cleaning up non-module level imports from any other files in `units`, which is beyond the scope of this pull request (I have not checked how much work that would be).

Reply:
It would definitely be good to separate them out into a `.errors.py` module.